jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c

```diff
@@ -574,396 +472,262 @@ typedef enum
+    {
+      arg_int--;
+    }
 
-  JERRY_ASSERT (p - buff < buffer_size);
-  /* String terminator. */
-  *p = 0;
-  ecma_string_t *str = ecma_new_ecma_string_from_utf8 (buff, (lit_utf8_size_t) (p - buff));
+    ecma_stringbuilder_append_char (&builder, LIT_CHAR_0);
 
-  ret_value = ecma_make_string_value (str);
-  JMEM_FINALIZE_LOCAL_ARRAY (buff);
+    if (arg_int > 0)
+    {
+      ecma_stringbuilder_append_char (&builder, LIT_CHAR_DOT);
+    }
 
-  return ret_value;
-} /* ecma_builtin_number_prototype_object_to_fixed */
+    for (int32_t i = 0; i < arg_int; i++)
+    {
+      ecma_stringbuilder_append_char (&builder, LIT_CHAR_0);
+    }
 
-/**
- * The Number.prototype object's 'toExponential' routine
- *
- * See also:
- *          ECMA-262 v5, 15.7.4.6
- *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
- */
-static ecma_value_t
-ecma_builtin_number_prototype_object_to_exponential (ecma_number_t this_num, /**< this argument number */
-                                                     ecma_value_t fraction_digits) /**< routine's argument */
-{
-  bool is_negative;
-  int32_t frac_digits;
+    if (mode == NUMBER_ROUTINE_TO_EXPONENTIAL)
+    {
+      ecma_stringbuilder_append_raw (&builder, (const lit_utf8_byte_t *) "e+0", 3);
+    }
 
-  ecma_value_t comp_value = ecma_builtin_number_prepare_conversion (&this_num,
-                                                                    fraction_digits,
-                                                                    &is_negative,
-                                                                    &frac_digits,
-                                                                    NUMBER_ROUTINE_TO_EXPONENTIAL);
+    return ecma_make_string_value (ecma_stringbuilder_finalize (&builder));
+  }
 
-  if (!ecma_is_value_empty (comp_value))
+  /* Handle infinity separately */
+  if (ecma_number_is_infinity (this_num))
   {
-    return comp_value;
+    ecma_stringbuilder_append_magic (&builder, LIT_MAGIC_STRING_INFINITY_UL);
+    return ecma_make_string_value (ecma_stringbuilder_finalize (&builder));
   }
 
-  /* Get the parameters of the number if non zero. */
-  lit_utf8_byte_t digits[ECMA_MAX_CHARS_IN_STRINGIFIED_NUMBER];
-  lit_utf8_size_t num_digits;
-  int32_t exponent;
+  num_of_digits = ecma_number_to_decimal (this_num, digits, &exponent);
 
-  if (!ecma_number_is_zero (this_num))
+  /* Handle undefined argument */
+  if (ecma_is_value_undefined (arg) && mode == NUMBER_ROUTINE_TO_EXPONENTIAL)
   {
-    num_digits = ecma_number_to_decimal (this_num, digits, &exponent);
-  }
-  else
-  {
-    digits[0] = '0';
-    num_digits = 1;
-    exponent = 1;
+    arg_int = (int32_t) num_of_digits - 1;
   }
 
-  if (ecma_is_value_undefined (fraction_digits))
+  if (mode == NUMBER_ROUTINE_TO_FIXED
+      && exponent > 21)
   {
-    frac_digits = (int32_t) num_digits - 1;
+    ecma_value_t result_num = ecma_builtin_number_prototype_object_to_string (this_num, NULL, 0);
```

Comment:
How about destroying the builder first, and returning with the result of the to_string call?